python_modules/dagster/dagster/_core/definitions/source_asset.py

```diff
@@ -24,27 +27,44 @@
     DEFAULT_IO_MANAGER_KEY,
     validate_group_name,
 )
-from dagster._core.errors import DagsterInvalidDefinitionError, DagsterInvalidInvocationError
+from dagster._core.errors import (
+    DagsterInvalidDefinitionError,
+    DagsterInvalidInvocationError,
+    DagsterInvalidObservationError,
+)
+from dagster._core.execution.context.compute import OpExecutionContext
 from dagster._core.storage.io_manager import IOManagerDefinition
 from dagster._utils import merge_dicts
 from dagster._utils.backcompat import ExperimentalWarning, experimental_arg_warning
 
+if TYPE_CHECKING:
+    from dagster._core.execution.context.compute import SourceAssetObserveContext
 
-class SourceAsset(
-    NamedTuple(
-        "_SourceAsset",
-        [
-            ("key", PublicAttr[AssetKey]),
-            ("metadata_entries", Sequence[Union[MetadataEntry, PartitionMetadataEntry]]),
-            ("io_manager_key", PublicAttr[Optional[str]]),
-            ("description", PublicAttr[Optional[str]]),
-            ("partitions_def", PublicAttr[Optional[PartitionsDefinition]]),
-            ("group_name", PublicAttr[str]),
-            ("resource_defs", PublicAttr[Dict[str, ResourceDefinition]]),
-        ],
-    ),
-    ResourceAddable,
-):
+
+class SourceAssetObserveFunctionWithContext(Protocol):
```

Comment:
Why not use `Callable[[SourceAssetObserveContext], MetadataUserInput`?

Reply:
Because the function only optionally takes a context function. This can't be represented with `Callable` unfortunately (that's why we don't do it for regular assets).

Reply:
I was imagining:
```
SourceAssetObserveFunction: TypeAlias = Union[
    Callable[[SourceAssetObserveContext], MetadataUserInput], Callable[[], MetadataUserInput]
]

```

Reply:
Yeah, this was the first thing I tried and for some reason it does not work (though logically it should).

Reply:
got it - was not a blocking concern, just curious